src/gas-receiver/AxelarGasReceiver.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.9;
+
+import { IERC20 } from '../interfaces/IERC20.sol';
+import { IAxelarGasReceiver } from '../interfaces/IAxelarGasReceiver.sol';
+
+// This should be owned by the microservice that is paying for gas.
+contract AxelarGasReceiver is IAxelarGasReceiver {
+    error NotOwner();
+
+    // bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)
+    bytes32 internal constant _IMPLEMENTATION_SLOT = 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+    // keccak256('owner');
+    bytes32 internal constant _OWNER_SLOT = 0x02016836a56b71f0d02689e69e326f4f4c1b9057164ef592671cf0d37c8040c0;
+
+    modifier onlyOwner() {
+        if (owner() != msg.sender) revert NotOwner();
+        _;
+    }
+
+    function owner() public view returns (address owner_) {
+        assembly {
+            owner_ := sload(_OWNER_SLOT)
+        }
+    }
+
+    function setup(bytes calldata data) external override {
+        address newOwner = abi.decode(data, (address));
+        if (newOwner != address(0)) {
+            emit OwnershipTransferred(newOwner);
+            assembly {
+                sstore(_OWNER_SLOT, newOwner)
+            }
+        }
+    }
+
+    // This is called on the source chain before calling the gateway to execute a remote contract.
+    function payGasForContractCall(
+        address sender,
+        string memory destinationChain,
```

Comment:
In order to not reintroduce issues any c4 brought up, can we set all of these memory variables to calldata?